src/test/java/dev/blaauwendraad/masker/json/ValueMaskersTest.java

```diff
@@ -0,0 +1,206 @@
+
+        // checking that can be used with all types in the builder
+        KeyMaskingConfig.builder()
+                .maskStringsWith(valueMasker)
+                .maskNumbersWith(valueMasker)
+                .maskBooleansWith(valueMasker)
+                .build();
+
+        Assertions.assertThat(ByteValueMaskerContext.maskStringWith("\"secret\"", valueMasker)).isEqualTo("null");
+        Assertions.assertThat(ByteValueMaskerContext.maskNumberWith(12345, valueMasker)).isEqualTo("null");
+        Assertions.assertThat(ByteValueMaskerContext.maskBooleanWith(true, valueMasker)).isEqualTo("null");
+    }
+
+    @Test
+    void customStringValueMasker() {
+        ValueMasker.StringMasker valueMasker =
+                context -> context.replaceBytes(0, context.byteLength(), "null".getBytes(StandardCharsets.UTF_8), 1);
+
+        // checking that can be used with string types in the builder
+        KeyMaskingConfig.builder().maskStringsWith(valueMasker).build();
+
+        Assertions.assertThat(ByteValueMaskerContext.maskStringWith("\"secret\"", valueMasker)).isEqualTo("null");
+    }
+
+    @Test
+    void customNumberValueMasker() {
+        ValueMasker.NumberMasker valueMasker =
+                context -> context.replaceBytes(0, context.byteLength(), "null".getBytes(StandardCharsets.UTF_8), 1);
+
+        // checking that can be used with number types in the builder
+        KeyMaskingConfig.builder().maskNumbersWith(valueMasker).build();
+
+        Assertions.assertThat(ByteValueMaskerContext.maskNumberWith(12345, valueMasker)).isEqualTo("null");
+    }
+
+    @Test
+    void customBooleanValueMasker() {
+        ValueMasker.BooleanMasker valueMasker =
+                context -> context.replaceBytes(0, context.byteLength(), "null".getBytes(StandardCharsets.UTF_8), 1);
+
+        // checking that can be used with boolean types in the builder
+        KeyMaskingConfig.builder().maskBooleansWith(valueMasker).build();
+
+        Assertions.assertThat(ByteValueMaskerContext.maskBooleanWith(true, valueMasker)).isEqualTo("null");
+    }
+
+    @Test
+    void maskEverySecondCharacter() {
+        ValueMasker.StringMasker valueMasker =
+                context -> context.replaceBytes(0, context.byteLength(), everyOtherCharacterMasked(context), 1);
+
+        // checking that it can be used with string types in the builder
+        KeyMaskingConfig.builder().maskStringsWith(valueMasker).build();
+
+        Assertions.assertThat(ByteValueMaskerContext.maskStringWith("hello", valueMasker)).isEqualTo("h*l*o");
+    }
+
+    @Nonnull
+    private byte[] everyOtherCharacterMasked(ValueMaskerContext valueMaskerContext) {
+        byte[] maskResult = new byte[valueMaskerContext.byteLength()];
+        for (int i = 0; i < valueMaskerContext.byteLength(); i++) {
+            if (i % 2 == 0) {
+                maskResult[i] = valueMaskerContext.getByte(i);
+            } else {
+                maskResult[i] = '*';
+            }
+        }
```

Comment:
This implementation is slightly wrong - the string value must have been used with quotes (see other tests) and the implementation itself should have skipped the quotes, otherwise you might mask the closing quote with `*` producing an invalid JSON.

Reply:
Yikes, the fact that I didn't remember that one day after discussing is not a good sign

Reply:
I also was planning to write more end to end json masker like this instead of this unit actually. Should've done that...